[Feature] Publish models after training if  published_keys is set in CheckpointHook

KerwinKai opened the pull request:
## Motivation

As https://github.com/open-mmlab/mmengine/issues/905 described.

## Modification

[Enhancement] add automatically Publish in checkpointhook.py and corresponding update document in hook.md

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

For example：

I am training cifar10 using the resnet50 model from the code in the mmengine example docs. And add default_hook in runner

```
default_hooks = dict(checkpoint=dict(type='CheckpointHook', interval=1,
                                     save_best='accuracy', rule='less', published_keys=['meta', 'state_dict']))
runner = Runner(
    model=MMResNet50(),
    work_dir='./work_dir',
    train_dataloader=train_dataloader,
    optim_wrapper=dict(optimizer=dict(type=SGD, lr=0.001, momentum=0.9)),
    train_cfg=dict(by_epoch=True, max_epochs=5, val_interval=1),
    val_dataloader=val_dataloader,
    val_cfg=dict(),
    val_evaluator=dict(type=Accuracy),
    default_hooks=default_hooks,
)
```

After trainning will publish the model, here is logs:
```
Loads checkpoint by local backend from path: /data/run01/scz0b6e/mm-lab/run_place/work_dir/epoch_5.pth
03/08 15:55:18 - mmengine - INFO - Load checkpoint from /data/run01/scz0b6e/mm-lab/run_place/work_dir/epoch_5.pth
03/08 15:55:18 - mmengine - INFO - Key `message_hub` will be removed because it is not in save_keys. If you want to keep it, please set `message_hub` in published_keys
03/08 15:55:18 - mmengine - INFO - Key `optimizer` will be removed because it is not in save_keys. If you want to keep it, please set `optimizer` in published_keys
03/08 15:55:18 - mmengine - INFO - The published model is saved at /data/run01/scz0b6e/mm-lab/run_place/work_dir/epoch_5-d16194b7.pth.
Loads checkpoint by local backend from path: /data/run01/scz0b6e/mm-lab/run_place/work_dir/best_accuracy_epoch_1.pth
03/08 15:55:18 - mmengine - INFO - Load checkpoint from /data/run01/scz0b6e/mm-lab/run_place/work_dir/best_accuracy_epoch_1.pth
03/08 15:55:18 - mmengine - INFO - Key `message_hub` will be removed because it is not in save_keys. If you want to keep it, please set `message_hub` in published_keys
03/08 15:55:19 - mmengine - INFO - The published model is saved at /data/run01/scz0b6e/mm-lab/run_place/work_dir/best_accuracy_epoch_1-107ccaf7.pth.
```

Here is the resulting folder structure

```
(base) [scz0b6e@ln01 run_place]$ cd work_dir/
(base) [scz0b6e@ln01 work_dir]$ ll
total 1302396
drwxrwxr-x 3 scz0b6e scz0b6e      4096 Mar  8 15:50 20230308_155014
-rw-rw-r-- 1 scz0b6e scz0b6e         1 Mar  8 15:50 20230308_155014.py
-rw-rw-r-- 1 scz0b6e scz0b6e 102497337 Mar  8 15:55 best_accuracy_epoch_1-107ccaf7.pth
-rw-rw-r-- 1 scz0b6e scz0b6e 102670911 Mar  8 15:51 best_accuracy_epoch_1.pth
-rw-rw-r-- 1 scz0b6e scz0b6e 204935167 Mar  8 15:51 epoch_1.pth
-rw-rw-r-- 1 scz0b6e scz0b6e 205061629 Mar  8 15:52 epoch_2.pth
-rw-rw-r-- 1 scz0b6e scz0b6e 205186557 Mar  8 15:53 epoch_3.pth
-rw-rw-r-- 1 scz0b6e scz0b6e 205311485 Mar  8 15:54 epoch_4.pth
-rw-rw-r-- 1 scz0b6e scz0b6e 102497337 Mar  8 15:55 epoch_5-d16194b7.pth
-rw-rw-r-- 1 scz0b6e scz0b6e 205436221 Mar  8 15:55 epoch_5.pth
-rw-rw-r-- 1 scz0b6e scz0b6e        57 Mar  8 15:55 last_checkpoint
```

Use follow code to check and compare the keys
```
import torch
def process_checkpoint(in_file):
    print(in_file)
    checkpoint = torch.load(in_file, map_location='cpu')
    ckpt_keys = list(checkpoint.keys())
    print(ckpt_keys)
    print()

def main():
    process_checkpoint('/HOME/scz0b6e/run/mm-lab/run_place/work_dir/best_accuracy_epoch_1.pth')
    process_checkpoint('/HOME/scz0b6e/run/mm-lab/run_place/work_dir/best_accuracy_epoch_1-107ccaf7.pth')
    process_checkpoint('/HOME/scz0b6e/run/mm-lab/run_place/work_dir/epoch_5.pth')
    process_checkpoint('/HOME/scz0b6e/run/mm-lab/run_place/work_dir/epoch_5-d16194b7.pth')

if __name__ == '__main__':
    main()
```

And here is the log:
```
/HOME/scz0b6e/run/mm-lab/run_place/work_dir/best_accuracy_epoch_1.pth
['meta', 'state_dict', 'message_hub']

/HOME/scz0b6e/run/mm-lab/run_place/work_dir/best_accuracy_epoch_1-107ccaf7.pth
['meta', 'state_dict']

/HOME/scz0b6e/run/mm-lab/run_place/work_dir/epoch_5.pth
['meta', 'state_dict', 'message_hub', 'optimizer']

/HOME/scz0b6e/run/mm-lab/run_place/work_dir/epoch_5-d16194b7.pth
['meta', 'state_dict']
```
The full logs: https://github.com/KerwinKai/mmengine/blob/main/20230308_155014.log

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/open-mmlab/mmengine?pullRequest=987) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/open-mmlab/mmengine?pullRequest=987) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/open-mmlab/mmengine?pullRequest=987) it.</sub>

HAOCHENYE commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -218,6 +224,21 @@ def __init__(self,
                 self.best_ckpt_path_dict: Dict = dict()
 
+        # published keys
+        assert (isinstance(published_keys, str)
+                or is_list_of(published_keys, str)
+                or (published_keys is None)), (
+                    '"published_keys" should be a str or list of str or None, '
+                    f'but got {type(published_keys)}')
```
```suggestion
        if not (isinstance(published_keys, str)
                or is_list_of(published_keys, str)
                or (published_keys is None)), (
                    '"published_keys" should be a str or list of str or None, '
                    f'but got {type(published_keys)}'):
                    raise TypeError(...)
```

HAOCHENYE commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -218,6 +224,21 @@ def __init__(self,
+                    '"published_keys" should be a str or list of str or None, '
+                    f'but got {type(published_keys)}')
+
+        if isinstance(published_keys, list):
+            assert len(published_keys) == len(set(published_keys)), (
+                'Find duplicate element in "published_keys".')
+        else:
+            if published_keys is not None:
+                published_keys = [published_keys]
```
```suggestion
        if isinstace(published_keys, str):
            published_keys = [published_keys]    
        elif isinstance(published_keys, list):
            assert len(published_keys) == len(set(published_keys)), (
                'Find duplicate element in "published_keys".')
        self.published_keys = published_keys
```

HAOCHENYE commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +325,64 @@ def after_val_epoch(self, runner, metrics):
+        Args:
+            runner (Runner): The runner of the training process.
+        """
+        if not self.published_keys:
```
```suggestion
        if self.published_keys is None:
```

HAOCHENYE commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +325,64 @@ def after_val_epoch(self, runner, metrics):
+
+        if self.save_last:
+            last_ckpt = runner.message_hub.get_info('last_ckpt')
+            if not last_ckpt:
+                assert ('Did not find last_checkpoint to be resumed.')
```
What will happen if `last_ckpt` is not in `message_hub` when we call `runner.message_hub.get_info('last_ckpt')`
```suggestion
            assert last_ckpt ('Did not find last_checkpoint to be resumed.')
```

HAOCHENYE commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +325,64 @@ def after_val_epoch(self, runner, metrics):
+            self._publish_model(runner, last_ckpt)
+
+        if self.save_best is not None:
+            best_ckpt = self.best_ckpt_path
+            if not best_ckpt:
+                assert ('Did not find best_checkpoint to be resumed.')
```
```suggestion
            if not self.best_ckpt_path:
                raise RuntimeError(xxx)
```

HAOCHENYE commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +325,64 @@ def after_val_epoch(self, runner, metrics):
+            self._publish_model(runner, best_ckpt)
+
+    def _publish_model(self, runner, out_file) -> None:
+        import subprocess
+
+        import torch
```
import these modules at the beginning

HAOCHENYE commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +325,64 @@ def after_val_epoch(self, runner, metrics):
+                assert ('Did not find best_checkpoint to be resumed.')
+            self._publish_model(runner, best_ckpt)
+
+    def _publish_model(self, runner, out_file) -> None:
```
```suggestion
    def _publish_model(self, runner, out_file: str) -> None:
```

HAOCHENYE commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +325,64 @@ def after_val_epoch(self, runner, metrics):
+
+        import torch
+
+        checkpoint = runner.load_checkpoint(out_file)
+        ckpt_keys = list(checkpoint.keys())
+        published_keys = self.published_keys
+        if not isinstance(published_keys, list):
+            return
```
```suggestion
        if published_keys is None:
            return
        checkpoint = runner.load_checkpoint(out_file)
        ckpt_keys = list(checkpoint.keys())
        published_keys = self.published_keys
```

HAOCHENYE commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +325,64 @@ def after_val_epoch(self, runner, metrics):
+                    f'save_keys. If you want to keep it, '
+                    f'please set `{k}` in published_keys',
+                    logger='current')
+                checkpoint.pop(k, None)
```
Since k must be a key of `ckpt_keys`, we do not need to set `None` as the default pop value.
```suggestion
                checkpoint.pop(k)
```

HAOCHENYE commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +325,64 @@ def after_val_epoch(self, runner, metrics):
+                    f'please set `{k}` in published_keys',
+                    logger='current')
+                checkpoint.pop(k, None)
+        if out_file.endswith('.pth'):
+            out_file_name = out_file[:-4]
+        else:
+            out_file_name = out_file
```
Why do we need to make sure `tmp_out_file_name` ends with a suffix `.pth`?
```suggestion
        out_file_name = osp.splitext(out_file)[0]
```

HAOCHENYE commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +325,64 @@ def after_val_epoch(self, runner, metrics):
+        else:
+            out_file_name = out_file
+        tmp_out_file_name = out_file + '.pth'
+        if torch.__version__ >= '1.6':
+            torch.save(
+                checkpoint,
+                tmp_out_file_name,
+                _use_new_zipfile_serialization=False)
+        else:
+            torch.save(checkpoint, tmp_out_file_name)
```
MMEngine requires PyTorch version >= 1.6. We do not need to check it here.
```suggestion
        torch.save(checkpoint, tmp_out_file_name)
```

HAOCHENYE commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +325,64 @@ def after_val_epoch(self, runner, metrics):
+            torch.save(checkpoint, tmp_out_file_name)
+        sha = subprocess.check_output(['sha256sum',
+                                       tmp_out_file_name]).decode()
+        final_file = out_file_name + f'-{sha[:8]}.pth'
+        subprocess.Popen(['mv', tmp_out_file_name, final_file])
```
Is it necessary to use sha to identify the checkpoint (maybe a timestamp is enough)? If it is necessary, we can use `hashlib` the calculate this, and do not need to save it as a file before.
```suggestion
        final_file = out_file_name + f'-{sha[:8]}.pth'
        subprocess.Popen(['mv', tmp_out_file_name, final_file])
```

zhouzaida commented on docs/en/tutorials/hook.md:
```diff
@@ -120,6 +121,14 @@ The four features mentioned above are described below.
 
+- Automatically publish the best and the last checkpoints
+
+  If you want to publish the best and the last checkpoints after train, you can set the `published_keys` parameter. You can select any keys in checkpoint to be published.
```
```suggestion
  If you want to publish the best and the last checkpoints after training, you can set the `published_keys` parameter. You can select any keys in checkpoint to be published.
```

zhouzaida commented:
Hi, it would be better to also update `docs/zh_cn/tutorials/hook.md`.

zhouzaida commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -84,7 +86,9 @@ class CheckpointHook(Hook):
-
+        published_keys (str, List[str], optional): If ``save_last`` is ``True``
+            or ``save_best`` is not ``None``, it will automatically
+            publish model with keys in list after train. Defaults to None.
```
```suggestion
            publish model with keys in the list after training. Defaults to None.
```

zhouzaida commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -95,6 +99,9 @@ class CheckpointHook(Hook):
         >>> # Save best based on multi metrics with different comparison rule
         >>> CheckpointHook(interval=2, by_epoch=True,
         >>>                save_best=['FID', 'IS'], rule=['less', 'greater'])
+        >>> # Save best based on single metric and publish model after train
```
```suggestion
        >>> # Save best based on single metric and publish model after training
```

zhouzaida commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -218,6 +226,21 @@ def __init__(self,
                 self.best_ckpt_path_dict: Dict = dict()
 
+        # published keys
+        if not (isinstance(published_keys, str)
+                or is_list_of(published_keys, str)
+                or (published_keys is None)), (
+                    '"published_keys" should be a str or list of str or None, '
+                    f'but got {type(published_keys)}'):
+                    raise TypeError(...)
```
```suggestion
        if not (isinstance(published_keys, str)
                or is_list_of(published_keys, str)
                or published_keys is None):
            raise TypeError(
                '"published_keys" should be a str or list of str or None, '
                f'but got {type(published_keys)}'):
```

zhouzaida commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +327,50 @@ def after_val_epoch(self, runner, metrics):
         self._save_best_checkpoint(runner, metrics)
 
+    def after_train(self, runner) -> None:
+        """Publish the checkpoint after train epoch.
```
```suggestion
        """Publish the checkpoint after training.
```

zhouzaida commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +327,50 @@ def after_val_epoch(self, runner, metrics):
+            self._publish_model(runner, last_ckpt)
+
+        if self.save_best is not None:
+            if not self.best_ckpt_path:
+                raise RuntimeError(xxx)
+            self._publish_model(runner, best_ckpt)
```
There are two cases. The best checkpoints may have more than 1.

https://github.com/open-mmlab/mmengine/blob/12e3d4e1e122dcd3dd183460ab2dd733f16ef26a/mmengine/hooks/checkpoint_hook.py#L415-L423

zhouzaida commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +327,50 @@ def after_val_epoch(self, runner, metrics):
+        if self.published_keys is None:
+            return
+
+        if self.save_last:
+            last_ckpt = runner.message_hub.get_info('last_ckpt')
+            assert last_ckpt ('Did not find last_checkpoint to be resumed.')
+            self._publish_model(runner, last_ckpt)
```
```suggestion
        if self.save_last and 'last_ckpt' in self.runtime_info:
            last_ckpt = runner.message_hub.get_info('last_ckpt')
            self._publish_model(runner, last_ckpt)
```

zhouzaida commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +327,50 @@ def after_val_epoch(self, runner, metrics):
+
+        if published_keys is None:
+            return
+        checkpoint = runner.load_checkpoint(out_file)
```
`runner.load_checkpoint` will load the checkpoint to model which is not expected.
we can use `_load_checkpoint` with `from mmengine.runner.checkpoint import _load_checkpoint`

KerwinKai commented:
> Hi, it would be better to also update `docs/zh_cn/tutorials/hook.md`.

add it now~

zhouzaida commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +327,50 @@ def after_val_epoch(self, runner, metrics):
+        if published_keys is None:
+            return
+        checkpoint = runner.load_checkpoint(out_file)
+        ckpt_keys = list(checkpoint.keys())
+        published_keys = self.published_keys
+        for k in ckpt_keys:
+            if k not in published_keys:
+                print_log(
+                    f'Key `{k}` will be removed because it is not in '
+                    f'save_keys. If you want to keep it, '
+                    f'please set `{k}` in published_keys',
+                    logger='current')
+                checkpoint.pop(k)
```
```suggestion
        removed_keys = []
        for key in checkpoint.keys():
            if key not in self.published_keys:
                removed_keys.append(key)
                checkpoint.pop(k)
        if removed_keys:     
            print_log(
                f'Key {removed_keys} will be removed because they are not '
                'found in published_keys. If you want to keep them, '
                f'please set `{removed_keys}` in published_keys',
                logger='current')
```

HAOCHENYE commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +327,47 @@ def after_val_epoch(self, runner, metrics):
+                'found in published_keys. If you want to keep them, '
+                f'please set `{removed_keys}` in published_keys',
+                logger='current')
+        print(runner.timestamp)
```
Clean the code
```suggestion
        print(runner.timestamp)
```

HAOCHENYE commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +327,47 @@ def after_val_epoch(self, runner, metrics):
+                f'please set `{removed_keys}` in published_keys',
+                logger='current')
+        print(runner.timestamp)
+        final_file = osp.splitext(out_file)[0] + f'-{runner.timestamp}.pth'
```
```suggestion
        final_file = osp.splitext(out_file)[0] + f'-published-{runner.timestamp}.pth'
```

HAOCHENYE commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +327,47 @@ def after_val_epoch(self, runner, metrics):
+                logger='current')
+        print(runner.timestamp)
+        final_file = osp.splitext(out_file)[0] + f'-{runner.timestamp}.pth'
+        torch.save(checkpoint, final_file)
```
It could be better to use save_checkpoint(`from mmengine.runner import save_checkpoint`) to save the published model to make sure the checkpoint can be uploaded to non-local backend.

HAOCHENYE commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +327,47 @@ def after_val_epoch(self, runner, metrics):
+
+        checkpoint = _load_checkpoint(out_file)
+        removed_keys = []
+        for key in checkpoint.keys():
+            if key not in self.published_keys:
+                removed_keys.append(key)
+                checkpoint.pop(k)
```
Suggest creating a new `dict` to store the published keys. It could be dangerous to pop the item during transverse the `checkpoint`.
```suggestion
        for key in list(checkpoint.keys()):
            if key not in self.published_keys:
                removed_keys.append(key)
                checkpoint.pop(k)
```

HAOCHENYE commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +327,47 @@ def after_val_epoch(self, runner, metrics):
+        if self.published_keys is None:
+            return
+
+        if self.save_last and 'last_ckpt' in self.runtime_info:
```
where is `self.runtime_info` defined.

zhouzaida replied:
`if self.save_last and 'last_ckpt' in runner.message_hub.runtime_info:`

HAOCHENYE commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +327,47 @@ def after_val_epoch(self, runner, metrics):
+            last_ckpt = runner.message_hub.get_info('last_ckpt')
+            self._publish_model(runner, last_ckpt)
+
+        if self.save_best is not None:
+            if len(self.key_indicators) == 1:
+                if self.best_ckpt_path is not None:
+                    self._publish_model(runner, self.best_ckpt_path)
+            else:
+                for key, best_ckpt in self.best_ckpt_path_dict.items():
+                    self._publish_model(runner, best_ckpt)
```
Will this format more clear?
```suggestion
        if getattr(self, 'best_ckpt_path', None) is not None:
            self._publish_model(runner, self.best_ckpt_path)
        if getattr(self, 'best_ckpt_path_dict', None) is not None:
            for key, best_ckpt in self.best_ckpt_path_dict.items():
                self._publish_model(runner, best_ckpt)
```

HAOCHENYE commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -218,6 +226,21 @@ def __init__(self,
+            published_keys = [published_keys]    
+        elif isinstance(published_keys, list):
+            assert len(published_keys) == len(set(published_keys)), (
+                'Find duplicate element in "published_keys".')
```
```suggestion
                'Find duplicate elements in "published_keys".')
```

KerwinKai commented:
I have modified the code as suggested above, this is the latest running log and file generation, I will finish writing `test_checkpoint_hook.py` as soon as possible

log:
```
03/09 21:48:20 - mmengine - INFO - Epoch(val) [5][313/313]  accuracy: 51.0500
Loads checkpoint by local backend from path: /data/run01/scz0b6e/mm-lab/run_place/work_dir/epoch_5.pth
03/09 21:48:20 - mmengine - INFO - Key ['message_hub', 'optimizer'] will be removed because they are not found in published_keys. If you want to keep them, please set `['message_hub', 'optimizer']` in published_keys
03/09 21:48:20 - mmengine - WARNING - "FileClient" will be deprecated in future. Please use io functions in https://mmengine.readthedocs.io/en/latest/api/fileio.html#file-io
03/09 21:48:20 - mmengine - INFO - The published model is saved at /data/run01/scz0b6e/mm-lab/run_place/work_dir/epoch_5-published-20230309_214304.pth.
Loads checkpoint by local backend from path: /data/run01/scz0b6e/mm-lab/run_place/work_dir/best_accuracy_epoch_1.pth
03/09 21:48:21 - mmengine - INFO - Key ['message_hub'] will be removed because they are not found in published_keys. If you want to keep them, please set `['message_hub']` in published_keys
03/09 21:48:21 - mmengine - WARNING - "FileClient" will be deprecated in future. Please use io functions in https://mmengine.readthedocs.io/en/latest/api/fileio.html#file-io
03/09 21:48:21 - mmengine - INFO - The published model is saved at /data/run01/scz0b6e/mm-lab/run_place/work_dir/best_accuracy_epoch_1-published-20230309_214304.pth.
```

detail log: https://github.com/KerwinKai/mmengine/blob/main/20230309_214304.log

file tree
```
(base) [scz0b6e@ln01 work_dir]$ ll
total 1302504
drwxrwxr-x 3 scz0b6e scz0b6e      4096 Mar  9 21:43 20230309_214304
-rw-rw-r-- 1 scz0b6e scz0b6e         1 Mar  9 21:43 20230309_214304.py
-rw-rw-r-- 1 scz0b6e scz0b6e 102671423 Mar  9 21:44 best_accuracy_epoch_1.pth
-rw-rw-r-- 1 scz0b6e scz0b6e 102544001 Mar  9 21:48 best_accuracy_epoch_1-published-20230309_214304.pth
-rw-rw-r-- 1 scz0b6e scz0b6e 204935743 Mar  9 21:44 epoch_1.pth
-rw-rw-r-- 1 scz0b6e scz0b6e 205062461 Mar  9 21:45 epoch_2.pth
-rw-rw-r-- 1 scz0b6e scz0b6e 205188093 Mar  9 21:46 epoch_3.pth
-rw-rw-r-- 1 scz0b6e scz0b6e 205313917 Mar  9 21:47 epoch_4.pth
-rw-rw-r-- 1 scz0b6e scz0b6e 205440125 Mar  9 21:48 epoch_5.pth
-rw-rw-r-- 1 scz0b6e scz0b6e 102544001 Mar  9 21:48 epoch_5-published-20230309_214304.pth
-rw-rw-r-- 1 scz0b6e scz0b6e        57 Mar  9 21:48 last_checkpoint
```

zhouzaida commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +327,51 @@ def after_val_epoch(self, runner, metrics):
+            for key, best_ckpt in self.best_ckpt_path_dict.items():
+                self._publish_model(runner, best_ckpt)
+
+    def _publish_model(self, runner, out_file: str) -> None:
+        """Publish the checkpoint.
+
+        Args:
+            runner (Runner): The runner of the training process.
+            out_file (str): The checkpoint path that ought to be published.
+        """
+        checkpoint = _load_checkpoint(out_file)
+        published_keys = self.published_keys
+        removed_keys = []
+        for key in list(checkpoint.keys()):
+            if key not in published_keys:
+                removed_keys.append(key)
+                checkpoint.pop(key)
+        if removed_keys:
+            print_log(
+                f'Key {removed_keys} will be removed because they are not '
+                'found in published_keys. If you want to keep them, '
+                f'please set `{removed_keys}` in published_keys',
+                logger='current')
+        final_file = osp.splitext(
+            out_file)[0] + f'-published-{runner.timestamp}.pth'
+        save_checkpoint(checkpoint, final_file)
+        print_log(
+            f'The published model is saved at {final_file}.', logger='current')
```
```suggestion
    def _publish_model(self, runner, ckpt_path: str) -> None:
        """Remove unnecessary keys from ckpt_path and save the new checkpoint.

        Args:
            runner (Runner): The runner of the training process.
            ckpt_path (str): The checkpoint path that ought to be published.
        """
        checkpoint = _load_checkpoint(ckpt_path)
        published_keys = self.published_keys
        removed_keys = []
        for key in list(checkpoint.keys()):
            if key not in published_keys:
                removed_keys.append(key)
                checkpoint.pop(key)
        if removed_keys:
            print_log(
                f'Key {removed_keys} will be removed because they are not '
                'found in published_keys. If you want to keep them, '
                f'please set `{removed_keys}` in published_keys',
                logger='current')
        final_path = osp.splitext(
            ckpt_path)[0] + f'-published-{runner.timestamp}.pth'
        save_checkpoint(checkpoint, final_path)
        print_log(
            f'The published model is saved at {final_path}.', logger='current')
```

KerwinKai commented:
Hi, I've updated the test docs and I'm having two issues. @HAOCHENYE 
First, when I run 'test_checkpoint_hook.py'  with following code. The command line will prompt the following error. I think it is a problem with the order of package references. After testing, 'from mmengine.runner import save_checkpoint'
'from mmengine.runner.checkpoint import _load_checkpoint' placed under the ‘_publish_model‘ function can avoid this problem
'''
if __name__ == '__main__':
    test_ckpt = TestCheckpointHook()
    work_dir = 'work_dir'
    test_ckpt.test_with_runner(work_dir)
'''
‘’‘
ImportError: cannot import name 'Hook' from partially initialized module 'mmengine.hooks' (most likely due to a circular import)
‘’‘
Second, 'test_after_train(self, tmp_path)' is a bit hard to write, because its runtime 'load_checkpoint' requires a .pth file, which is generated by runner.train(). So I put the test content in the 'test_with_runner' function.
‘’‘
    def test_after_train(self, tmp_path):
        runner = Mock()
        work_dir = str(tmp_path)
        runner.work_dir = work_dir
        runner.epoch = 9
        runner.model = Mock()
        runner.message_hub = MessageHub.get_instance('test_after_train')
        
        publish_ckpt_names = 'publish_ckpt_names'

        # check best ckpt name and best score
        
        metrics = {'acc': 0.5, 'map': 0.3}
        checkpoint_hook = CheckpointHook(
            interval=2, by_epoch=True, save_best='acc', 
            rule='greater', published_keys=['meta', 'state_dict'])
        checkpoint_hook.before_train(runner)
        checkpoint_hook.after_train_epoch(runner)
        checkpoint_hook.after_val_epoch(runner, metrics)
        checkpoint_hook.after_train(runner)

if __name__ == '__main__':
    test_ckpt = TestCheckpointHook()
    work_dir = 'work_dir'
    test_ckpt.test_after_train(work_dir)
‘’‘

‘’‘
Loads checkpoint by locl backend from path: work_dir/epoch_10.pth
File "/home/run/mmengine/tests/test_hooks/test_checkpoint_hook.py", line 529, in <module>
    test_ckpt.test_after_train(work_dir)
  File "/home/run/mmengine/tests/test_hooks/test_checkpoint_hook.py", line 479, in test_after_train
    checkpoint_hook.after_train(runner)
  File "/home/run/mmengine/mmengine/hooks/checkpoint_hook.py", line 344, in after_train
    self._publish_model(runner, last_ckpt)
  File "/home/run/mmengine/mmengine/hooks/checkpoint_hook.py", line 361, in _publish_model
    checkpoint = _load_checkpoint(ckpt_path)
  File "/home/run/mmengine/mmengine/runner/checkpoint.py", line 518, in _load_checkpoint
    return CheckpointLoader.load_checkpoint(filename, map_location, logger)
  File "/home/run/mmengine/mmengine/runner/checkpoint.py", line 300, in load_checkpoint
    return checkpoint_loader(filename, map_location)
  File "/home/run/mmengine/mmengine/runner/checkpoint.py", line 316, in load_from_local
    raise FileNotFoundError(f'{filename} can not be found.')
FileNotFoundError: work_dir/epoch_10.pth can not be found.
‘’‘

Based on the consideration of the above two problems, I just updated the code a little bit, hoping to get some suggestions

KerwinKai commented:
Hi, I've updated the test docs and I'm having two issues. @HAOCHENYE
First, when I run `test_checkpoint_hook.py` with following code. The command line will prompt the following error. I think it is a problem with the order of package references. After testing, `from mmengine.runner import save_checkpoint`
`from mmengine.runner.checkpoint import _load_checkpoint`placed under the `_publish_model` function can avoid this problem
```
if name == 'main':
test_ckpt = TestCheckpointHook()
work_dir = 'work_dir'
test_ckpt.test_with_runner(work_dir)
```
```
ImportError: cannot import name 'Hook' from partially initialized module 'mmengine.hooks' (most likely due to a circular import)
```
Second, `test_after_train(self, tmp_path)` is a bit hard to write, because its runtime `load_checkpoint` requires a .pth file, which is generated by `runner.train()`. So I put the test content in the `test_with_runner` function.
```
   def test_after_train(self, tmp_path):
      runner = Mock()
      work_dir = str(tmp_path)
      runner.work_dir = work_dir
      runner.epoch = 9
      runner.model = Mock()
      runner.message_hub = MessageHub.get_instance('test_after_train')

    publish_ckpt_names = 'publish_ckpt_names'
    
    # check best ckpt name and best score
    
    metrics = {'acc': 0.5, 'map': 0.3}
    checkpoint_hook = CheckpointHook(
        interval=2, by_epoch=True, save_best='acc', 
        rule='greater', published_keys=['meta', 'state_dict'])
    checkpoint_hook.before_train(runner)
    checkpoint_hook.after_train_epoch(runner)
    checkpoint_hook.after_val_epoch(runner, metrics)
    checkpoint_hook.after_train(runner)
    
if name == 'main':
   test_ckpt = TestCheckpointHook()
   work_dir = 'work_dir'
   test_ckpt.test_after_train(work_dir)
```

```
Loads checkpoint by locl backend from path: work_dir/epoch_10.pth
File "/home/run/mmengine/tests/test_hooks/test_checkpoint_hook.py", line 529, in
test_ckpt.test_after_train(work_dir)
File "/home/run/mmengine/tests/test_hooks/test_checkpoint_hook.py", line 479, in test_after_train
checkpoint_hook.after_train(runner)
File "/home/run/mmengine/mmengine/hooks/checkpoint_hook.py", line 344, in after_train
self._publish_model(runner, last_ckpt)
File "/home/run/mmengine/mmengine/hooks/checkpoint_hook.py", line 361, in _publish_model
checkpoint = _load_checkpoint(ckpt_path)
File "/home/run/mmengine/mmengine/runner/checkpoint.py", line 518, in _load_checkpoint
return CheckpointLoader.load_checkpoint(filename, map_location, logger)
File "/home/run/mmengine/mmengine/runner/checkpoint.py", line 300, in load_checkpoint
return checkpoint_loader(filename, map_location)
File "/home/run/mmengine/mmengine/runner/checkpoint.py", line 316, in load_from_local
raise FileNotFoundError(f'{filename} can not be found.')
FileNotFoundError: work_dir/epoch_10.pth can not be found.
```

Based on the consideration of the above two problems, I just updated the code a little bit, hoping to get some suggestions

KerwinKai commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +325,61 @@ def after_val_epoch(self, runner, metrics):
+        published_keys = self.published_keys
+        removed_keys = []
+        for key in list(checkpoint.keys()):
+            if published_keys is not None and\
```
In order to avoid mypy error in pre-commit (mmengine\hooks\checkpoint_hook.py:360: error: Unsupported right operand type for in ("Optional[List[str]]")), I have to judge again that published_keys is not None .

HAOCHENYE commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +325,61 @@ def after_val_epoch(self, runner, metrics):
+        from mmengine.runner import save_checkpoint
+        from mmengine.runner.checkpoint import _load_checkpoint
+        checkpoint = _load_checkpoint(ckpt_path)
+        published_keys = self.published_keys
```
Hi, you can pre-define the type of published_keys to pass mypy check
```suggestion
        published_keys: Union[list, str] = self.published_keys
```

HAOCHENYE commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +325,61 @@ def after_val_epoch(self, runner, metrics):
+        else:
+            ckpt_path_list = runner.message_hub.get_info('publish_ckpt_names')
+            ckpt_path_list.append(final_path)
+            runner.message_hub.update_info('publish_ckpt_names',
+                                           ckpt_path_list)
```
Is this line necessary?

KerwinKai replied:
just for test check. I will change it

KerwinKai replied:
If I rm this. I could get 'lask_ckpt' from 'runner.message_hub.get_info('last_ckpt')', but how could I check the 'best_ckpt' when it is a dict. I could not use 'self.best_ckpt_path_dict.items()' in 'test_with_runner'.

KerwinKai replied:
After test, it is not necessary. rm it, and the message_hub still save 'ckpt_path_list'

zhouzaida commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +325,59 @@ def after_val_epoch(self, runner, metrics):
+        from mmengine.runner import save_checkpoint
+        from mmengine.runner.checkpoint import _load_checkpoint
+        checkpoint = _load_checkpoint(ckpt_path)
+        published_keys: Optional[Union[List[str], str]] = self.published_keys
+        removed_keys = []
+        for key in list(checkpoint.keys()):
+            if published_keys is not None and\
+                    key not in published_keys:
+                removed_keys.append(key)
+                checkpoint.pop(key)
```
```suggestion
        assert self.published_keys is not None
        removed_keys = []
        for key in list(checkpoint.keys()):
            if key not in self.published_keys:
                removed_keys.append(key)
                checkpoint.pop(key)
```

zhouzaida commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +325,58 @@ def after_val_epoch(self, runner, metrics):
+        save_checkpoint(checkpoint, final_path)
+        print_log(
+            f'The published model is saved at {final_path}.', logger='current')
+        if 'publish_ckpt_names' not in runner.message_hub.runtime_info:
+            runner.message_hub.update_info('publish_ckpt_names', [final_path])
+        else:
+            ckpt_path_list = runner.message_hub.get_info('publish_ckpt_names')
+            ckpt_path_list.append(final_path)
```
Is it necessary to add those messages to message_hub?

KerwinKai replied:
For check in test_checkpoint_hook.py, because I could load self.best_ckpt_path in it

HAOCHENYE replied:
We should not modify the code logic just for writing unit tests. A more reasonable approach is to check the checkpoint of the published model has been generated.

KerwinKai replied:
> We should not modify the code logic just for writing unit tests. A more reasonable approach is to check the checkpoint of the published model has been saved.

oki, already rm the unnecessary messages add in message_hub.

zhouzaida commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +325,56 @@ def after_val_epoch(self, runner, metrics):
+                f'Key {removed_keys} will be removed because they are not '
+                'found in published_keys. If you want to keep them, '
+                f'please set `{removed_keys}` in published_keys',
+                logger='current')
```
Hi, we can follow the implementation from https://github.com/open-mmlab/mmdetection/blob/892bcdad0cb7f693c4e2745e4d187b21dd1fddae/tools/model_converters/publish_model.py#L44 to call `subprocess.check_output(['sha256sum', out_file]).decode()`.

zhouzaida commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +325,56 @@ def after_val_epoch(self, runner, metrics):
+                logger='current')
+        final_path = osp.splitext(
+            ckpt_path)[0] + f'-published-{runner.timestamp}.pth'
+        for key in list(checkpoint.keys()):
+            assert key in list(self.published_keys)
```
This for loop seems like it can be removed because the following snippet does the same thing.

```python
for key in list(checkpoint.keys()):
    if key not in self.published_keys:
```

KerwinKai replied:
rm it now, for check in test_checkpoint_hook before

zhouzaida commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +325,56 @@ def after_val_epoch(self, runner, metrics):
+        for key in list(checkpoint.keys()):
+            assert key in list(self.published_keys)
+        save_checkpoint(checkpoint, final_path)
+        assert self.file_client.isfile(final_path)
```
Is it necessary to assert it is a file?

zhouzaida commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +325,56 @@ def after_val_epoch(self, runner, metrics):
+        save_checkpoint(checkpoint, final_path)
+        assert self.file_client.isfile(final_path)
+        print_log(
+            f'The published model is saved at {final_path}.', logger='current')
```
```suggestion
            f'The published model ({ckpt_path}) is saved at {final_path}.', logger='current')
```

zhouzaida commented on docs/zh_cn/tutorials/hook.md:
```diff
@@ -71,6 +71,7 @@ runner.train()
 - 保存最新的多个权重
 - 保存最优权重
 - 指定保存权重的路径
+- 自动发布最好的和最后的权重
```
```suggestion
- 制作发布用的权重
```

zhouzaida commented on docs/zh_cn/tutorials/hook.md:
````diff
@@ -121,6 +122,14 @@ runner.train()
   default_hooks = dict(checkpoint=dict(type='CheckpointHook', interval=5, out_dir='/path/of/directory'))
   ```
 
+- 自动发布最好的和最后的权重
+
+  如果你想在训练后发布最好的和最后的权重，你可以设置`published_keys`参数。 通过这个参数，您可以选择要发布权重中所要保存的键。
````
```suggestion
- 制作发布用的权重

  如果你想在训练结束后自动生成可发布的权重（删除不需要的权重，例如优化器状态），你可以设置 `published_keys` 参数，选择需要保留的信息。注意：需要相应设置 save_best 或者 save_last 参数，这样才会生成可发布的权重，其中设置 save_best 会生成最优权重的可发布权重，设置 save_last 会生成最后一个权重的可发布权重，这两个参数也可同时设置。
```

zhouzaida commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -84,7 +86,10 @@ class CheckpointHook(Hook):
+        published_keys (str, List[str], optional): If ``save_last`` is ``True``
+            or ``save_best`` is not ``None``, it will automatically
+            publish model with keys in the list after training.
+            Defaults to None.
```
```suggestion
            Defaults to None.
            `New in version 0.7.1.`
```

HAOCHENYE replied:
```suggestion
            Defaults to None.
            New in version ...
```

zhouzaida commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +327,56 @@ def after_val_epoch(self, runner, metrics):
+                logger='current')
+        checkpoint_data = pickle.dumps(checkpoint)
+        sha = hashlib.sha256(checkpoint_data).hexdigest()
+        final_path = osp.splitext(ckpt_path)[0] + f'-published-{sha[:8]}.pth'
```
```suggestion
        final_path = osp.splitext(ckpt_path)[0] + f'-{sha[:8]}.pth'
```

zhouzaida commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -304,6 +327,56 @@ def after_val_epoch(self, runner, metrics):
+        final_path = osp.splitext(ckpt_path)[0] + f'-published-{sha[:8]}.pth'
+        save_checkpoint(checkpoint, final_path)
+        print_log(
+            f'The published model ({ckpt_path}) is saved at '
+            f'{final_path}.',
```
```suggestion
            f'The checkpoint ({ckpt_path}) is published to '
            f'{final_path}.',
```

zhouzaida commented on docs/en/tutorials/hook.md:
```diff
@@ -120,6 +121,14 @@ The four features mentioned above are described below.
 
+- Make checkpoints for publish
+
+  If you want to automatically generate publishable checkpoints after training (remove unnecessary keys, such as optimizer state), you can set the `published_keys` parameter to choose which information to keep. Note: You need to set the `save_best` or `save_last` parameters accordingly, so that the releasable checkpoints will be generated. Setting `save_best` will generate the releasable weights  ...
```
```suggestion
  If you want to automatically generate publishable checkpoints after training (remove unnecessary keys, such as optimizer state), you can set the `published_keys` parameter to choose which information to keep. Note: You need to set the `save_best` or `save_last` parameters accordingly so that the releasable checkpoints will be generated. Setting `save_best` will generate the releasable weights of the optimal checkpoint, and setting `save_last` will generate the releasable final checkpoint. These two parameters can also be set at the same time.
```

HAOCHENYE commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -218,6 +227,20 @@ def __init__(self,
 
+        # published keys
+        if not (isinstance(published_keys, str)
+                or is_list_of(published_keys, str) or published_keys is None):
```
```suggestion
                or is_seq_of(published_keys, str) or published_keys is None):
```

HAOCHENYE commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -218,6 +227,20 @@ def __init__(self,
+        if not (isinstance(published_keys, str)
+                or is_list_of(published_keys, str) or published_keys is None):
+            raise TypeError(
+                '"published_keys" should be a str or list of str or None, '
```
```suggestion
                '"published_keys" should be a str or a sequence of str or None, '
```

HAOCHENYE commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -218,6 +227,20 @@ def __init__(self,
+
+        if isinstance(published_keys, str):
+            published_keys = [published_keys]
+        elif isinstance(published_keys, list):
```
```suggestion
        elif isinstance(published_keys, (list, tuple)):
```

zhouzaida commented on mmengine/hooks/checkpoint_hook.py:
```diff
@@ -218,6 +228,20 @@ def __init__(self,
+        if not (isinstance(published_keys, str)
+                or is_seq_of(published_keys, str) or published_keys is None):
+            raise TypeError(
+                '"published_keys" should be a str or a sequence of str or None, '
+                f'but got {type(published_keys)}')
```
```suggestion
                '"published_keys" should be a str or a sequence of str or '
                f'None, but got {type(published_keys)}')
```
